django/core/checks/security/base.py

```diff
@@ -235,6 +239,65 @@ def check_referrer_policy(app_configs, **kwargs):
     return []
 
 
+@register(Tags.security, deploy=True)
+def check_cache_and_media_root(app_configs, **kwargs):
```

Comment:
I'm not sure we need three separate functions. They're very similar and could be combined, something like this:

```python
@register(Tags.security, deploy=True)
def check_cache_location_not_exposed(app_configs, **kwargs):
    errors = []
    for name in ('MEDIA_ROOT', 'STATIC_ROOT', 'STATICFILES_DIRS'):
        setting = getattr(settings, name, None)
        if not setting:
            continue
        if name == 'STATICFILES_DIRS':
            paths = {pathlib.Path(dir).resolve() for dir in setting}
        else:
            paths = {pathlib.Path(setting).resolve()}
        for alias in settings.CACHES:
            cache = caches[alias]
            if not isinstance(cache, FileBaseCache):
                continue
            cache_path = pathlib.Path(cache._dir)
            if any(path in cache_path.parents for path in paths):
                errors.append(Warning(
                    "Your cache configuration might expose or corrupt your file system.",
                    hint=f"Your cache '{alias}' is inside your '{name}'",
                    id='cache.W002',
                ))
    return errors
```

Note the use of `pathlib` and the containment check that is possible with `.parents`. Super nice. :grinning:

Reply:
Two nested loops, if-elses and two continues. @pope1ni this might be a matter of personal preference 🤣 But that `pathlib` suggestion I can agree with 👍

Reply:
Two loops isn't so bad when the alternative is three nearly identical functions to maintain! 😄

There are relatively easy ways to remove the if-else and the first continue if you'd really like me to suggest an alternative?

Reply:
I found that `cache_path.parent` is not include `cache_path` and my solution is use 

```python
    cache_path = pathlib.Path(cache._dir).resolve()
    if any(path in cache_path.parents or path == cache_path for path in paths):
        ... 
```

to instead

```python
    cache_path = pathlib.Path(cache._dir)
    if any(path in cache_path.parents for path in paths):
        ... 
``` 

or there are more effective solution?

Reply:
Good point. We need to also check when they are the same.